jina/flow/base.py

```diff
@@ -605,6 +616,83 @@ def gather_inspect(
             # no inspect node is in the graph, return the current graph
             return self
 
+    def _get_gateway_target(self, prefix):
+        gateway_pod = self._pod_nodes[GATEWAY_NAME]
+        return (
+            f'{prefix}-{GATEWAY_NAME}',
+            {
+                'host': gateway_pod.head_host,
+                'port': gateway_pod.head_port_in,
+                'expected_parts': 0,
+            },
+        )
+
+    def _update_port_type(self, pod):
+        pod.head_args.socket_in = SocketType.ROUTER_BIND
+        pod.tail_args.dynamic_out_routing = True
+        pod.tail_args.socket_out = SocketType.DEALER_CONNECT
+
+    def _get_routing_graph(self):
+        target_pods = [
+            (
+                pod_id,
+                {
+                    'host': pod.head_host,
+                    'port': pod.head_port_in,
+                    'expected_parts': 0,
+                },
+            )
+            for pod_id, pod in self._pod_nodes.items()
+            if pod_id != GATEWAY_NAME
+        ]
+        target_pods = (
+            [self._get_gateway_target('start')]
+            + target_pods
+            + [self._get_gateway_target('end')]
+        )
+
+        pods_id_to_position = {
+            pod_id: (position, pod)
+            for position, (pod_id, pod) in enumerate(target_pods)
+        }
+
+        edges = defaultdict(list)
+        for end, pod in self._pod_nodes.items():
+            self._update_port_type(pod)
+            if end == GATEWAY_NAME:
+                end = f'end-{GATEWAY_NAME}'
+
+            for start in pod.needs:
+                if start == GATEWAY_NAME:
+                    start = f'start-{GATEWAY_NAME}'
+
+                if start not in pods_id_to_position:
+                    raise FlowMissingPodError(
+                        f'{start} is not in this flow, misspelled name?'
+                    )
+
+                end_pod = pods_id_to_position[end]
+                end_pod[1]['expected_parts'] += 1
+                start_pos = pods_id_to_position[start][0]
+                end_pos = end_pod[0]
+                edges[start_pos].append(end_pos)
+
+        targets = [pod[1] for pod in target_pods]
+        parsed_edges = {key: {'targets': value} for key, value in edges.items()}
+        graph = {'active_pod_index': 0, 'pods': targets, 'edges': parsed_edges}
+
+        routing_pb = jina_pb2.RoutingGraphProto()
+        json_format.ParseDict(graph, routing_pb)
+        return RoutingGraph(routing_pb)
+
+    def _init_dynamic_routing(self):
```

Comment:
I would name `_set_initial_dynamic_routing_graph`. This naming seems that is an initialization that will be changed.

Why I guess it is the `graph` that will be used by default by the `Gateway` right?